src/wallet/wallet.cpp

```diff
@@ -3007,37 +2999,34 @@ bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CTransac
 void CWallet::CommitTransaction(CTransactionRef tx, mapValue_t mapValue, std::vector<std::pair<std::string, std::string>> orderForm)
 {
     LOCK(cs_wallet);
-
-    CWalletTx wtxNew(this, std::move(tx));
-    wtxNew.mapValue = std::move(mapValue);
-    wtxNew.vOrderForm = std::move(orderForm);
-    wtxNew.fTimeReceivedIsTxTime = true;
-    wtxNew.fFromMe = true;
-
-    WalletLogPrintf("CommitTransaction:\n%s", wtxNew.tx->ToString()); /* Continued */
+    WalletLogPrintf("CommitTransaction:\n%s", tx->ToString()); /* Continued */
 
     // Add tx to wallet, because if it has change it's also ours,
     // otherwise just for transaction history.
-    AddToWallet(wtxNew);
+    CWalletTx* wtx = AddToWallet(std::move(tx), {}, [&](CWalletTx& wtx, bool new_tx) {
+        CHECK_NONFATAL(wtx.mapValue.empty());
+        CHECK_NONFATAL(wtx.vOrderForm.empty());
+        wtx.mapValue = std::move(mapValue);
```

Comment:
Does the `std::move` still work after going through a lambda capture by reference?

Reply:
Yes. mapValue is a reference to the same thing it was outside of the lambda. std::move casts it to an rvalue reference.

Reply:
Oh, this has nothing to do with lambdas. I didn't know that references (as long as they are not const) can be moved.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/9381#discussion_r420187394

> Oh, this has nothing to do with lambdas. I didn't know that references (as long as they are not const) can be moved.

Yep, `std::move` isn't affected by outside things like this. `std::move(mapValue)` is only a type cast from `mapValue_t&` to `mapValue_t&&` that makes the compiler favor the `operator=(mapValue_t&&)` assignment overload instead of the `operator=(const mapValue_t&)` one for `wtx.mapValue =`

Reply:
Even const ones, actually - though unless they have mutable fields, there is no effect.

Reply:
With "no effect" you mean `const&` will stay `const&`, right? See also the example, which does not compile, because the `const&` copy constructor is deleted:

```cpp
#include <memory>

struct Test{
    const std::unique_ptr<int>b;
    Test(const std::unique_ptr<int>& a):b{std::move(a)}{}
};

Reply:
`std::move(a)` where a is `const T&`, returns something of type `const T&&`. If T has for example a `operator=(const T&&)`, then that one would be selected in assignment. If there is no such operator, `operator=(const T&` will be selected instead. In general a `const T&&` assignment/constructor operator only makes sense if a class has mutable fields.

Reply:
TIL that `const T&&` exists.